Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.mm

```diff
@@ -112,191 +132,161 @@
+    : m_element(element), m_device(device), m_direction(dir), m_index(index),
+      m_min(IOHIDElementGetLogicalMin(m_element)), m_max(IOHIDElementGetLogicalMax(m_element))
 {
-  switch (dir)
-  {
-  case up:
-    m_name = "Up";
-    break;
-  case right:
-    m_name = "Right";
-    break;
-  case down:
-    m_name = "Down";
-    break;
-  case left:
-    m_name = "Left";
-    break;
-  default:
-    m_name = "unk";
-  }
 }
 
 ControlState Joystick::Hat::GetState() const
 {
   IOHIDValueRef value;
 
-  if (IOHIDDeviceGetValue(m_device, m_element, &value) == kIOReturnSuccess)
-  {
-    int position = IOHIDValueGetIntegerValue(value);
-    int min = IOHIDElementGetLogicalMin(m_element);
-    int max = IOHIDElementGetLogicalMax(m_element);
+  if (IOHIDDeviceGetValue(m_device, m_element, &value) != kIOReturnSuccess)
+    return 0;
 
-    // if the position is outside the min or max, don't register it as a valid button press
-    if (position < min || position > max)
-    {
-      return 0;
-    }
+  int position = IOHIDValueGetIntegerValue(value);
 
-    // normalize the position so that its lowest value is 0
-    position -= min;
+  // if the position is outside the min or max, don't register it as a valid button press
+  if (position < m_min || position > m_max)
+    return 0;
 
-    switch (position)
-    {
-    case 0:
-      if (m_direction == up)
-        return 1;
-      break;
-    case 1:
-      if (m_direction == up || m_direction == right)
-        return 1;
-      break;
-    case 2:
-      if (m_direction == right)
-        return 1;
-      break;
-    case 3:
-      if (m_direction == right || m_direction == down)
-        return 1;
-      break;
-    case 4:
-      if (m_direction == down)
-        return 1;
-      break;
-    case 5:
-      if (m_direction == down || m_direction == left)
-        return 1;
-      break;
-    case 6:
-      if (m_direction == left)
-        return 1;
-      break;
-    case 7:
-      if (m_direction == left || m_direction == up)
-        return 1;
-      break;
-    };
-  }
+  // normalize the position so that its lowest value is 0
+  position -= m_min;
 
-  return 0;
+  switch (position)
+  {
+  case 0:
+    if (m_direction == Direction::Up)
+      return 1;
+    break;
+  case 1:
+    if (m_direction == Direction::Up || m_direction == Direction::Right)
+      return 1;
+    break;
+  case 2:
+    if (m_direction == Direction::Right)
+      return 1;
+    break;
+  case 3:
+    if (m_direction == Direction::Right || m_direction == Direction::Down)
+      return 1;
+    break;
+  case 4:
+    if (m_direction == Direction::Down)
+      return 1;
+    break;
+  case 5:
+    if (m_direction == Direction::Down || m_direction == Direction::Left)
+      return 1;
+    break;
+  case 6:
+    if (m_direction == Direction::Left)
+      return 1;
+    break;
+  case 7:
+    if (m_direction == Direction::Left || m_direction == Direction::Up)
+      return 1;
+    break;
+  default:
+    return 0;
+  };
```

Comment:
1. that `;` at the end seems superfluous.
2. the method as a whole has no `return` statement at the end right now; so any case match that doesn't `return 1` will not return any well-defined result (due to moving the previous unxonditional `return` into the `default` case).

Reply:
Good call. I'll just eliminate the whole switch statement with some math.

Reply:
I'm kinda excited and terrified at the same time :D
Guessing you'll try and compare `m_direction` (more or less) directly to `position`?

Reply:
Done. :P

Reply:
Hats off (no pun intended) at that math...lets hope people more mathy than me understand it 😛